### PR TITLE
GMX-05

### PR DESCRIPTION
**Pull request by ericmartihaynes**
# GMX-05

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [x] New Bot
- [ ] Fix a bot error
- [ ] Update a bot logic/test/documentation
- [ ] Code style update (formatting, renaming)
- [ ] Other (please describe):

## New Bot PR readiness

- [x] `npm install` command works (i.e. package-lock is updated properly)
- [x] The `name` and `description` fields of `package.json` describe the bot properly
- [x] All the libraries listed in `package.json` are being used
- [x] All the development libraries are in `devDependencies` section of `package.json`
- [x] The bot README is updated
- [x] The bot README sufficiently describes what the bot does
- [x] All the alerts that the bot emits are listed in the README
- [x] The txn/block examples listed in the README generate the expected findings
- [x] Code is easy to follow and/or properly commented
- [x] Bot contains tests
- [x] All tests pass when running `npm test`
- [x] Tests are not making real network calls (i.e. network interaction is properly mocked)
- [x] Tests cover all the paths (i.e. negative & positive test cases)
- [x] When the tests complete, no warning/errors are shown
- [x] `npm start` works without errors when executed using a jsonRpcUrl for each of the networks specified in the README
- [x] The networks specified in the README and the networks are listed in the `package.json` are the same
- [x] The code does what the README describes
- [x] Network calls are properly being cached (if duplicated calls can occur)
- [ ] A performance review has been done (e.g. `Promise.all` is used whenever possible)
- [x] The findings fields are filled out correctly
- [x] Findings `protocol` is specified (if the bot is for a specific one)
- [x] `metadata` of the findings contains all the information needed.
- [x] The SDK methods are being used appropriately (e.g. `filterLog`, `filterFunction`, `getEthersProvider`)
- [x] `ABI`s match the contracts used / `args` used from `LogDescriptions` exists in the `ABI`s
- [ ] The bot follows Forta recommended [best practices](https://docs.forta.network/en/latest/best-practices/)
- [x] Bot can recover from failed error calls when they can fail (e.g. errors are being caught on calls that can fail)
- [x] `repository` is added in the `package.json`

## Comments (optional)

If there is some useful information or important note for reviewing/understanding this bot please describe it here.

### CODE REVIEW

File: GMX-Bots/gmx5-detect-frontruns/src/agent.ts [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+          else {
+            callHistory.set(accountBack + token0 + token1, [swapEvent, callHistoryIndex, txEvent.hash]);
+            callHistoryKeys[callHistoryIndex] = accountBack + token0 + token1;
+          }
+        }
+        //transaction gets stored
+        else {
+          callHistory.set(accountBack + token0 + token1, [swapEvent, callHistoryIndex, txEvent.hash]);
+          callHistoryKeys[callHistoryIndex] = accountBack + token0 + token1;
+        }
```
Comment (e1Ru1o) [inline GMX-Bots/gmx5-detect-frontruns/src/agent.ts:137]:
```suggestion
        //transaction gets stored
        if (!sandwichAttack) {
          callHistory.set(accountBack + token0 + token1, [swapEvent, callHistoryIndex, txEvent.hash]);
          callHistoryKeys[callHistoryIndex] = accountBack + token0 + token1;
        }
```

File: GMX-Bots/gmx5-detect-frontruns/src/agent.ts [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+      });
+    }
+    return findings;
+  };
+
+// const handleBlock: HandleBlock = async (blockEvent: BlockEvent) => {
+//   const findings: Finding[] = [];
+//   // detect some block condition
+//   return findings;
+// }
```
Comment (e1Ru1o) [inline GMX-Bots/gmx5-detect-frontruns/src/agent.ts:199]: remove this

File: GMX-Bots/gmx5-detect-frontruns/src/agent.ts [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+export default {
+  handleTransaction: provideHandleTx(GMX_ROUTER_ADDRESS, SWAP_EVENT, getEthersProvider()),
+  // handleBlock
```
Comment (e1Ru1o) [inline GMX-Bots/gmx5-detect-frontruns/src/agent.ts:203]: remove this

File: GMX-Bots/gmx5-detect-frontruns/src/agent.spec.ts [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+import { Interface } from "@ethersproject/abi";
+
+const ABI: string[] = [
+  "event Swap(address account, address tokenIn, address tokenOut, uint256 amountIn, uint256 amountOut)",
+];
+
+const TEST_IFACE: Interface = new Interface([
+  ...ABI,
+  "event Swap(address account, address tokenIn, address tokenOut, uint256 amountIn, uint256 amountOut)",
+]);
```
Comment (e1Ru1o) [inline GMX-Bots/gmx5-detect-frontruns/src/agent.spec.ts:23]:
```suggestion
const TEST_IFACE: Interface = new Interface(ABI);
```

File: GMX-Bots/gmx5-detect-frontruns/src/agent.spec.ts [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+  "event Swap(address account, address tokenIn, address tokenOut, uint256 amountIn, uint256 amountOut)";
+const handler = provideHandleTx(
+  MOCK_GMX_ROUTER_ADDRESS,
+  MOCK_SWAP_EVENT,
+  mockProvider as unknown as ethers.providers.Provider
+);
+
+describe("sandwich attack frontrun agent", () => {
+  let handleTransaction: HandleTransaction;
+  const mockTxEvent = createTransactionEvent({} as any);
```
Comment (e1Ru1o) [inline GMX-Bots/gmx5-detect-frontruns/src/agent.spec.ts:37]: remove this because it is not used

**Review by e1Ru1o — CHANGES_REQUESTED** [review]

File: GMX-Bots/gmx5-detect-frontruns/.gitattributes [hunk 6]
```diff
@@ -0,0 +1,2 @@
+# Auto detect text files and perform LF normalization
```
[the remark this answers is not in the document]
Comment (ericmartihaynes) [reply GMX-Bots/gmx5-detect-frontruns/.gitattributes:1]: No idea how it got there, removed

File: GMX-Bots/gmx5-detect-frontruns/README.md [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+## Changing Network
+By default this bot uses the Arbitrum network, but it can be changed to Avalanche by setting "jsonRpcUrl" and "traceRpcUrl" to
+https://api.avax.network/ext/bc/C/rpc in the forta.config.json file. The GMX_ROUTER_ADDRESS in agent.ts must be changed to "0x5F719c2F1095F7B9fc68a68e35B51194f4b6abe8".
+
+## Test Data
+
+The bot behaviour can be verified with the following transactions:
+
+- ?????????????????????????????
```
[the remark this answers is not in the document]
Comment (ericmartihaynes) [reply GMX-Bots/gmx5-detect-frontruns/README.md:30]: I don't have any transactions to test with, should I just remove this section?

File: GMX-Bots/gmx5-detect-frontruns/src/agent.ts [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+          );
+        }
+      });
+    }
+    return findings;
+  };
+
+export default {
+  initialize: initialize(provider),
+  handleTransaction: provideHandleTx(GMX_ROUTER_ADDRESS, SWAP_EVENT, provider),
```
Comment (e1Ru1o) [inline GMX-Bots/gmx5-detect-frontruns/src/agent.ts:204]: this is not going to work, please do the following experiment:

1. Add a `console.log` in the `initialize` function to display the address you get from the `NetworkManager`.
2. Add a `console.log` in the line 50 for example to display `router`.
3. Do `npm run` on one of the supported networks and see what is printed.

Reach me on slack after that to discuss

**Review by e1Ru1o — CHANGES_REQUESTED** [review]

File: GMX-Bots/gmx5-detect-frontruns/src/agent.ts [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+              if (
+                (event[1] > potentialSandwhichFront[1] &&
+                  event[1] < callHistoryIndex &&
+                  callHistoryIndex > potentialSandwhichFront[1]) ||
+                (event[1] < potentialSandwhichFront[1] &&
+                  event[1] < callHistoryIndex &&
+                  callHistoryIndex < potentialSandwhichFront[1]) ||
+                (event[1] > potentialSandwhichFront[1] &&
+                  event[1] > callHistoryIndex &&
+                  callHistoryIndex < potentialSandwhichFront[1])
```
Comment (yasmine9) [inline GMX-Bots/gmx5-detect-frontruns/src/agent.ts:121]: I don't get what are you exactly checking in here. Can you explain ? 
I tried to understand this from a logic pov and i found out that: 

1.  **The first check:** `(event[1] > potentialSandwhichFront[1] &&  event[1] < callHistoryIndex &&  callHistoryIndex > potentialSandwhichFront[1])` is simply equivalent to 
`event[1] > potentialSandwhichFront[1] &&  event[1] < callHistoryIndex`. ie. the 3rd condition is not needed because 1+2 implies 3. 

2. **The second check:** `(event[1] < potentialSandwhichFront[1] && event[1] < callHistoryIndex && callHistoryIndex < potentialSandwhichFront[1])` is equivalent to 
`event[1] < callHistoryIndex && callHistoryIndex < potentialSandwhichFront[1]`. ie. the first condition is not needed because 2+3=> 1 . 
   
3. **The third check:**  is equivalent to `event[1] > potentialSandwhichFront[1] && callHistoryIndex < potentialSandwhichFront[1]`. ie the 2nd condition is not needed. 

Now overall it seems like you are checking if : 

```
potentialSandwhichFront[1] < event[1] < callHistoryIndex ||
event[1] < callHistoryIndex < potentialSandwhichFront[1] ||
callHistoryIndex < potentialSandwhichFront[1] < event[1]
```
I don't think this is what you intended to do, can you please explain this part.

File: GMX-Bots/gmx5-detect-frontruns/src/agent.ts [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+                  victimTokenIn = tokenInVictim;
+                  victimTokenOut = tokenOutVictim;
+                  victimTokenInAmount = amountInVictim;
+                  victimTokenOutAmount = amountOutVictim;
+                  frontrunnerAddress = accountFront;
+                  frontrunnerProfit = amountOutBack - amountInFront;
+                  victimAddressAndTokens = addressAndTokens;
+                  frontrunnerAddressAndTokens = accountBack + token0 + token1;
+                }
+              }
```
Comment (yasmine9) [inline GMX-Bots/gmx5-detect-frontruns/src/agent.ts:138]: Why you decided to store the values then check again if there's a frontRun later. 

It would be more organized to have something like this pattern:
```
if ( victim condition checked) {
// set sandwichAttack to true
// delete both victim and frontrunner from history 
// create finding using the the event args. ( no need to store them somewhere else) 
}
```

File: GMX-Bots/gmx5-detect-frontruns/src/agent.ts [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+          }
+        }
+
+        if (!sandwichAttack) {
+          callHistory.set(accountBack + token0 + token1, [swapEvent, callHistoryIndex, txEvent.hash]);
+          callHistoryKeys[callHistoryIndex] = accountBack + token0 + token1;
+        }
+
+        callHistoryIndex++;
+        mapCleanerCounter++;
```
Comment (yasmine9) [inline GMX-Bots/gmx5-detect-frontruns/src/agent.ts:149]: Why the Index is incremented even if a sandwichAttack is detected?

File: GMX-Bots/gmx5-detect-frontruns/src/agent.ts [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+};
+let GMX_ROUTER_ADDRESS = { value: "" };
+const provider = getEthersProvider();
+export const initialize = (provider: ethers.providers.Provider) => async () => {
+  const networkManager = new NetworkManager(data, (await provider.getNetwork()).chainId);
+  GMX_ROUTER_ADDRESS.value = networkManager.get("address");
+};
+export const SWAP_EVENT =
+  "event Swap(address account, address tokenIn, address tokenOut, uint256 amountIn, uint256 amountOut)";
+
```
Comment (yasmine9) [inline GMX-Bots/gmx5-detect-frontruns/src/agent.ts:37]: I'd recommend moving the network data, ABIs and some constants ( `CALL_HISTORY_SIZE` for example) to a separate file.

**Review by yasmine9 — CHANGES_REQUESTED** [review]

File: GMX-Bots/gmx5-detect-frontruns/src/agent.ts [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+  HandleTransaction,
+  TransactionEvent,
+  FindingSeverity,
+  FindingType,
+  Log,
+  LogDescription,
+  ethers,
+  getEthersProvider,
+} from "forta-agent";
+import { NetworkManager } from "forta-agent-tools";
```
[the remark this answers is not in the document]
Comment (ericmartihaynes) [reply GMX-Bots/gmx5-detect-frontruns/src/agent.ts:10]: Removed theProvider. As for the `networkManager`, I had it the way you suggested before, but had to change it because it was passed to `provideHandleTx` by value and not by reference, so the changes that happened in `initialize` didn't do anything. Lázaro suggested wrapping it in an object as a solution to this problem

File: GMX-Bots/gmx5-detect-frontruns/src/agent.ts (hunk 9 above)
Comment (ericmartihaynes) [reply GMX-Bots/gmx5-detect-frontruns/src/agent.ts:117]: Some of the conditions are redundant like you pointed out, but I thought it made more sense to leave them like this for the sake of understanding. Basically this condition checks if the victim transaction is in the sandwich, but because we have an array of a limited size that is similar to a circular array, we can't just check if the index of the victim transaction is in between the front and the back one (because some transaction could find the array full and occupy an earlier index). 
So there are 3 possible situations, where in each `potentialSandwhichFront[1]` is the front transaction, `event[1]` is the victim transaction, and `callHistoryIndex` is the back transaction:

1. In the first case (assuming array size 20), we check that the victim transaction is after the front one, the victim one is before the back one, and the back one is after the front one. Some example values can be `4, 6, 9`
2. In the second case (assuming array size 20), we check that the victim transaction is before the front transaction, the victim one is before the back one, and the back one is before the front. Some example values can be `18, 10, 15`
3. In the third case (assuming array size 20), we check that the victim transaction is after the front transaction, the victim one is after the back transaction, and the back one is before the front one. Some example values can be `10, 17, 5`

To sum up: Considering a circular structure (where the element after the last one is the first one), the transactions must be in some permutation that is in ascending order

File: GMX-Bots/gmx5-detect-frontruns/src/agent.ts (hunk 10 above)
Comment (ericmartihaynes) [reply GMX-Bots/gmx5-detect-frontruns/src/agent.ts:134]: I did it this way because of certain operations with the cleaner that need to occur before the victim and frontrunner are deleted from history. I tried to change it and move the cleaner to another part, but it was causing errors. I think in this case it might not be worth the time to change, but for next bots I will try to use a more clear pattern like the one you suggested

File: GMX-Bots/gmx5-detect-frontruns/src/agent.ts (hunk 11 above)
Comment (ericmartihaynes) [reply GMX-Bots/gmx5-detect-frontruns/src/agent.ts]: Causes no real problems but should not be incremented in this situation, fixed it

File: GMX-Bots/gmx5-detect-frontruns/src/agent.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (yasmine9) [reply GMX-Bots/gmx5-detect-frontruns/src/agent.ts:10]: What you had before is the `router` as a parameter which will be passed by value as you said. 
My suggesting is passing the `networkManager` as a parameter to `initialize` and `provideHandleTx`.

**Review by yasmine9 — COMMENTED** [review]
LGTM overall 👍. Final notes to be approved: 

- Some unused variables still have to be removed from the bot tests file. What I could see is import `ethers` from `forta-agent` and `mockProvider` variable ( consider removing the import as well). Check if any other are not used. 

- Please update the package-lock.json. this is an old version. 

- I left a comment concerning the router address discussion.

File: GMX-Bots/gmx5-detect-frontruns/src/agent.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (ericmartihaynes) [reply GMX-Bots/gmx5-detect-frontruns/src/agent.ts:10]: Okay, I understand what you mean now. The issue is if I have `networkManager` as a global variable, I can't assign it its value outside `initialize` or `provideHandleTx` (because there is an async call). So then I would have to change it's value when `initialize` is run. But wouldn't I run into the same issue, where `networkManager` will be passed to `provideHandleTx` by value?

[the remark this answers is not in the document]
Comment (yasmine9) [reply GMX-Bots/gmx5-detect-frontruns/src/agent.ts:10]: Yes, in that case you'll have the same problem. 

But you can avoid the async call when you create a new `NetworkManager` object, by passing only `util.data` as a parameter. Then initializing the chainId later by calling `init` function inside `initialize`.  You'll have something like: 

```
 const networkManager = new NetworkManager(util.data)

export const initialize = (provider: ethers.providers.Provider) => async () => {
await networkManager.init(provider)
...
}
```

**Review by yasmine9 — APPROVED** [review]

**Comment by ericmartihaynes** [comment]
@e1Ru1o Please approve this PR when you can

File: GMX-Bots/gmx5-detect-frontruns/package.json [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+  },
+  "devDependencies": {
+    "@types/jest": "^27.0.1",
+    "@types/nodemon": "^1.19.0",
+    "jest": "^27.0.6",
+    "nodemon": "^2.0.8",
+    "prettier": "^2.6.2",
+    "ts-jest": "^27.0.3",
+    "typescript": "^4.3.4",
+    "forta-agent-tools": "^2.0.7"
```
Comment (e1Ru1o) [inline GMX-Bots/gmx5-detect-frontruns/package.json:38]: this should be in `dependencies` this package is used in the `agent.ts`

**Review by e1Ru1o — CHANGES_REQUESTED** [review]
Ready to merge after this change

**Review by 0xtaf — APPROVED** [review]
